source/slang/slang-ir-diff-jvp.cpp

```diff
@@ -3,10 +3,303 @@
+            IRParam* paramD = as<IRParam>(differentiateParam(builder, paramP));
+            mapDifferentialInst(paramP, paramD);
+            newParamListD.add(paramD);
+        }
+
+        return newParamListD;
+    }
+
+    IRInst* differentiateVar(IRBuilder* builder, IRVar* varP)
+    {
+        if (IRType* typeD = differentiateType(builder, varP->getDataType()->getValueType()))
+        {
+            IRVar* varD = builder->emitVar(typeD);
+            SLANG_ASSERT(varD);
+            return varD;
+        }
+        return nullptr;
+    }
+
+    IRInst* differentiateBinaryArith(IRBuilder* builder, IRInst* arith)
+    {
+        SLANG_ASSERT(arith->getOperandCount() == 2);
+        
+        auto leftP = arith->getOperand(0);
+        auto rightP = arith->getOperand(1);
+
+        auto leftD = getDifferentialInst(leftP);
+        auto rightD = getDifferentialInst(rightP);
+
+        auto leftZero = builder->getFloatValue(leftP->getDataType(), 0.0);
+        auto rightZero = builder->getFloatValue(rightP->getDataType(), 0.0);
+
+        if (leftD || rightD)
+        {
+            leftD = leftD ? leftD : leftZero;
+            rightD = rightD ? rightD : rightZero;
+
+            // Might have to do special-case handling for non-scalar types,
+            // like float3 or float3x3
+            // 
+            auto resultType = arith->getDataType();
+            switch(arith->getOp())
+            {
+            case kIROp_Add:
+                return builder->emitAdd(resultType, leftD, rightD);
+            case kIROp_Mul:
+                return builder->emitAdd(resultType,
+                    builder->emitMul(resultType, leftD, rightP),
+                    builder->emitMul(resultType, leftP, rightD));
+            case kIROp_Sub:
+                return builder->emitSub(resultType, leftD, rightD);
+            case kIROp_Div:
+                return builder->emitDiv(resultType, 
+                    builder->emitSub(
+                        resultType,
+                        builder->emitMul(resultType, leftD, rightP),
+                        builder->emitMul(resultType, leftP, rightD)),
+                    builder->emitMul(
+                        rightP->getDataType(), rightP, rightP
+                    ));
+            default:
+                SLANG_UNEXPECTED("Attempting to differentiate unsupported arithmetic");
+            }
+        }
+
+        return nullptr;
+    }
+
+    IRInst* differentiateLoad(IRBuilder* builder, IRLoad* loadP)
+    {
+        if (auto varP = as<IRVar>(loadP->getPtr()))
+        {   
+            // If the loaded parameter has a differential version, 
+            // emit a load instruction for the differential parameter.
+            // Otherwise, emit nothing since there's nothing to load.
+            // 
+            if (auto varD = as<IRVar>(getDifferentialInst(varP)))
+            {
+                IRLoad* loadD = as<IRLoad>(builder->emitLoad(varD));
+                SLANG_ASSERT(loadD);
```

Comment:
I'd prefer to turn all `ASSERT`, `UNEXPECTED` etc. into diagnostics since that will inform the user what exactly is causing the compiler to fail.